src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectConfigurationStateSynchronizer.cs

```diff
@@ -234,4 +197,129 @@ public void ProjectConfigurationFileChanged(ProjectConfigurationFileChangeEventA
+
+    private void EnqueueProjectUpdate(ProjectKey projectKey, RazorProjectInfo? projectInfo, CancellationToken cancellationToken)
+    {
+        // Note: We lock on _projectUpdates to protect access to both _projectUpdates and UpdateItem.
+        lock (_projectUpdates)
+        {
+            if (!_projectUpdates.TryGetValue(projectKey, out var updateItem))
+            {
+                updateItem = new(projectKey);
+                _projectUpdates.Add(projectKey, updateItem);
+            }
+
+            // Ensure that we use the most recent project info, which could be null
+            // in the case of a reset.
+            updateItem.ProjectInfo = projectInfo;
+
+            // If this is new or its task has completed, start a new task to update
+            // the project after a delay. Note that we don't provide the project info
+            // because it might be updated before the delay completes.
+            if (updateItem.Task is null || updateItem.Task.IsCompleted)
+            {
+                updateItem.Task = UpdateAfterDelayAsync(projectKey, cancellationToken);
+            }
+        }
+
+        async Task UpdateAfterDelayAsync(ProjectKey projectKey, CancellationToken cancellationToken)
+        {
+            await Task.Delay(_delay, cancellationToken).ConfigureAwait(false);
+
+            // First, retrieve the project info (if any) from the update item.
+            RazorProjectInfo? projectInfo;
+
+            lock (_projectUpdates)
+            {
+                if (!_projectUpdates.TryGetValue(projectKey, out var updateItem))
+                {
+                    // The update item should not be removed before we've handled it here.
+                    Assumed.Unreachable();
+                    return;
+                }
+
+                projectInfo = updateItem.ProjectInfo;
+            }
+
+            await UpdateProjectAsync(projectKey, projectInfo, cancellationToken).ConfigureAwait(false);
+
+            // Now that the project update is complete, we can remove the update item, but
+            // be sure to protect access.
+            lock (_projectUpdates)
+            {
+                _projectUpdates.Remove(projectKey);
+            }
+
+            Task UpdateProjectAsync(ProjectKey projectKey, RazorProjectInfo? projectInfo, CancellationToken cancellationToken)
+            {
+                if (projectInfo is null)
+                {
+                    // When we're passed a null RazorProjectInfo, we reset the project back to an empty, unconfigured state.
+                    return _projectService.UpdateProjectAsync(
+                        projectKey,
+                        configuration: null,
+                        rootNamespace: null,
+                        displayName: "",
+                        ProjectWorkspaceState.Default,
+                        documents: [],
+                        cancellationToken);
```

Comment:
Nit: I would've factored this out into a separate ResetProjectAsync (I think original code had that?) that would also log the fact of project reset/removal (just as we log project update below).

Reply:
Yeah, I had that too originally, but it just had one caller, so I inlined it to reduce the number of local methods. Also, I feel like the concept of "resetting" a project is better represented by the new `EnqueueProjectReset` method which is declared alongside `EnqueueProjectUpdate`. This spot, which is a doubly nested local method is really more an implementation detail.

Reply:
Fair enough. I only commented because it was a little visually confusing to have to identically looking (though of course with different set of params) calls next to each other and didn't notice the "return" in front immediately. Logging might still make sense since we log the update - but not a reset? Don't feel strongly about either, feel free to punt.

Reply:
We log in `IRazorProjectService.UpdateProjectAsync(...)`, and I didn't want to add additional logging that might potentially pollute the logs:

https://github.com/dotnet/razor/blob/8558b71b160557aec3185720a06138f285ebc0c0/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectSystem/RazorProjectService.cs#L303-L326

Note that we didn't log for resets prior to the batching work queue change either.